src/metrics_probes.erl

```diff
@@ -0,0 +1,144 @@
+%%==============================================================================
+%% Copyright 2019 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
+-module(metrics_probes).
+-behaviour(exometer_probe).
+
+-include("mongoose_logger.hrl").
+
+%% exometer_entry callbacks
+%% exometer_probe callbacks
+-export([
+         behaviour/0,
+         probe_init/3,
+         probe_terminate/1,
+         probe_get_value/2,
+         probe_get_datapoints/1,
+         probe_update/2,
+         probe_reset/1,
+         probe_sample/1,
+         probe_setopts/3,
+         probe_handle_msg/2,
+         probe_code_change/3
+        ]).
+
+%% Samples
+-export([
+         do_sample_queue_lengths/0,
+         do_sample_tcp_connections/0
+        ]).
+
+-type datapoint() :: atom().
+-record(state, {
+    datapoints = [] :: [datapoint()],
+    data = #{} :: #{datapoint() => integer()},
+    sampling :: atom(),
+    ref :: reference() | undefined
+}).
+
+-spec behaviour() -> exometer:behaviour().
+behaviour() ->
+    probe.
+
+probe_init(_Name, _Type, Opts) ->
+    FunSampling = proplists:get_value(sampling, Opts, error_no_sampling_given),
+    DataPoints = proplists:get_value(datapoints, Opts, error_no_datapoints_given),
```

Comment:
Why won't we crash instead? This code enables us to enter incorrect state.

Reply:
I thought about that for a while, that's why I have weird default values, to see them clearly in the logs. Also, `exometer_probe` will call `probe_sample` as soon as it returns from init, so the crash will come very soon.
How should I make it crash here? `case of badfun -> throw`? I thought of that being defensive, dunno, tell me your opinion or some other idea please :)

Reply:
I'm not sure how to tackle it, but I would really suggest somehow crashing here. Exometer in it's API doesn't allow this function to return anything other then `{ok, any()}`, so I guess we need to `error(no_datapoints)` etc.

Also, sampling function could a be function reference here, not an arbitrary atom.